Crash in WebPageProxy::sendWheelEvent if the WebProcess takes too long to start

mcatanzaro opened the pull request:
#### 1e6c25ac7d906fe67c99b68697f787cbbdfedfe1
<pre>
Crash in WebPageProxy::sendWheelEvent if the WebProcess takes too long to start
<a href="https://bugs.webkit.org/show_bug.cgi?id=283546">https://bugs.webkit.org/show_bug.cgi?id=283546</a>

Reviewed by Adrian Perez de Castro.

Third time is the charm! We have previously landed two failed attempts
to fix frequent crashing in WebPageProxy::sendWheelEvent. Turns out
Lauro Moura had already debugged this and proposed a fix, which had
never landed. An AuxiliaryProcessProxy does not initially have a valid
IPC::Connection; instead it stores IPC messages into a queue, to be
processed and sent when the auxiliary process has finished launching. We
simply need to use that functionality instead of trying to access the
connection manually.

This commit is a rebased version of Lauro&apos;s work, except I use
WebPageProxy::sendToProcessContainingFrame to avoid breaking site
isolation, and have also changed
WebPageProxy::sendWheelEventScrollingAccelerationCurveIfNecessary, which
has the same bug. Finally, I also revert the unnecessary changes that I
previously landed in 302030@main.

Canonical link: <a href="https://commits.webkit.org/303867@main">https://commits.webkit.org/303867@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5c739c8e736d8122f98f8e7a9ba12052bb640088

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133835 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6345 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45025 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141412 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/85895 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/155/builds/6874 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/156/builds/6210 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/102380 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/85895 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136782 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/155/builds/6874 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/119982 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83178 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/155/builds/6874 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/macOS-Tahoe-Debug-API-Tests-EWS "Waiting to run tests") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/155/builds/6874 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144058 "Built successfully") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/153/builds/6016 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/38679 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/110750 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/160/builds/6098 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/156/builds/6210 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110946 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28136 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/116237 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/59763 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/158/builds/6068 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/34521 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/5913 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69532 "Built successfully") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/6159 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/6022 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -4295,20 +4295,17 @@ void WebPageProxy::sendWheelEvent(WebCore::FrameIdentifier frameID, const WebWhe
     Ref process = processContainingFrame(frameID);
     if (protectedDrawingArea()->shouldSendWheelEventsToEventDispatcher()) {
         sendWheelEventScrollingAccelerationCurveIfNecessary(frameID, event);
-        process->protectedConnection()->send(Messages::EventDispatcher::WheelEvent(webPageIDInProcess(process), event, rubberBandableEdges), 0, { }, Thread::QOS::UserInteractive);
```
Maybe the reason to use connection directly was to be able to pass the QOS?

cdumez replied:
I think this is a very good guess. Odds are we want to use `WebProcessProxy::send()` (for the queueing behavior when the connection is not established yet) and pipe the QoS through (if not already supported).

mcatanzaro replied:
No, AuxiliaryProcessProxy uses `QoS::UserInteractive` for all messages. So that's the default QoS.

mcatanzaro replied:
Specifically, in `AuxiliaryProcessProxy::didFinishLaunching` it calls `IPC::Connection::createServerConnection` using `Thread::QOS::UserInteractive`. So that QOS gets used as the default QOS.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -4295,20 +4295,17 @@ void WebPageProxy::sendWheelEvent(WebCore::FrameIdentifier frameID, const WebWhe
     if (protectedDrawingArea()->shouldSendWheelEventsToEventDispatcher()) {
         sendWheelEventScrollingAccelerationCurveIfNecessary(frameID, event);
-        process->protectedConnection()->send(Messages::EventDispatcher::WheelEvent(webPageIDInProcess(process), event, rubberBandableEdges), 0, { }, Thread::QOS::UserInteractive);
+        send(Messages::EventDispatcher::WheelEvent(webPageIDInProcess(process), event, rubberBandableEdges));
```
cc @achristensen07 @beidson @szewai Doesn't this send the IPC to a potentially different process now when site isolation is enabled? It seems it would always send to the main frame process now instead of the process backing the given frameID.

I'm no expert on site isolation but this looks very suspicious to me.

cdumez replied:
Can't we use `process->send()`? it would still do the queuing when the connection is not established yet.

mcatanzaro replied:
Good catch. `sendMouseEvent` and `sendKeyEvent` both use `sendToProcessContainingFrame`, and the frame is available, so we need to use that here too.

mcatanzaro replied:
Fixed.

mcatanzaro committed:
Crash in WebPageProxy::sendWheelEvent if the WebProcess takes too long to start
https://bugs.webkit.org/show_bug.cgi?id=283546

Reviewed by Adrian Perez de Castro.

Third time is the charm! We have previously landed two failed attempts
to fix frequent crashing in WebPageProxy::sendWheelEvent. Turns out
Lauro Moura had already debugged this and proposed a fix, which had
never landed. An AuxiliaryProcessProxy does not initially have a valid
IPC::Connection; instead it stores IPC messages into a queue, to be
processed and sent when the auxiliary process has finished launching. We
simply need to use that functionality instead of trying to access the
connection manually.

This commit is a rebased version of Lauro's work, except I use
WebPageProxy::sendToProcessContainingFrame to avoid breaking site
isolation, and have also changed
WebPageProxy::sendWheelEventScrollingAccelerationCurveIfNecessary, which
has the same bug. Finally, I also revert the unnecessary changes that I
previously landed in 302030@main.

Canonical link: https://commits.webkit.org/303867@main
